### PR TITLE
[ML] Tag destination index with data frame metadata

### PR DESCRIPTION
**Pull request by przemekwitek**
Tag all the destination indices that are created by Data Frame Analytics with metadata: created by, creation time, version and analytics id.

Also:
- extract destination index creation logic into a separate class: DataFrameAnalyticsIndex
- add unit test for DataframeIndex::createDestinationIndex method

### CODE REVIEW

**Comment by przemekwitek** [comment]
run elasticsearch-ci/docbldesx

File: x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/persistence/DataframeIndexTests.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.elasticsearch.action.admin.indices.create.CreateIndexAction;
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
+import org.elasticsearch.action.admin.indices.create.CreateIndexResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.xpack.core.dataframe.transforms.DataFrameTransformConfig;
+import org.elasticsearch.xpack.core.dataframe.transforms.DataFrameTransformConfigTests;
+import org.elasticsearch.xpack.core.watcher.watch.ClockMock;
```
Comment (droberts195) [inline x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/persistence/DataframeIndexTests.java]: I don't think you need to import this Watcher class, and it looks out-of-place to be importing another plugin's classes without a particularly good reason.  The only method you call on it is `fixed()`, and that is inherited from `java.time.Clock`.  So you might as well just replace `ClockMock.fixed()` with `Clock.fixed()` and delete this import.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsIndexTests.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsConfig;
+import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsDest;
+import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsSource;
+import org.elasticsearch.xpack.core.ml.dataframe.analyses.OutlierDetection;
+import org.elasticsearch.xpack.core.watcher.watch.ClockMock;
```
Comment (droberts195) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsIndexTests.java]: I don't think you need to import this Watcher class, and it looks out-of-place to be importing another plugin's classes without a particularly good reason.  The only method you call on it is `fixed()`, and that is inherited from `java.time.Clock`.  So you might as well just replace `ClockMock.fixed()` with `Clock.fixed()` and delete this import.

**Review by droberts195 — APPROVED** [review]
LGTM, although I think you can very simply remove a couple of dubious dependencies before merging

File: x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/persistence/DataframeIndexTests.java (hunk 1 above)
Comment (przemekwitek) [reply x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/persistence/DataframeIndexTests.java]: Done.
I missed the fact "fixed" was inherited. Thanks for pointing that out!

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsIndexTests.java (hunk 2 above)
Comment (przemekwitek) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsIndexTests.java]: Done.

**Comment by przemekwitek** [comment]
run elasticsearch-ci/2
run elasticsearch-ci/bwc
run elasticsearch-ci/default-distro

**Pull request merged**